File: src/main/scala/scala_examples/4_functional_composition.scala
package scala_examples

object FunctionalComposition extends App {

  // --------------------------------- Part 1 - Composition of a greeting ---------------------------------
  println("------------ Part 1 ------------\n")

  val addGreeting = (name: String) => s"Hello, my name is $name"
  val toUpperCase = (text: String) => text.toUpperCase

  // Scala supports compose and andThen to combine functions, the only difference is the order of applying
  val loudGreeting = addGreeting.andThen(toUpperCase)
  print(loudGreeting("Tom"))


  // --------------------------------- Part 2 - Passing functions around ---------------------------------
  println("\n------------ Part 2 ------------\n")

  // First we create a generic divisible_by function
  // Notice divisibleBy is a function that returns another function
  val divisibleBy = (y: Int) => (x: Int) => x % y == 0

  // We can now create a divisibleBy2 function using divisibleBy
  val divisibleBy2 = divisibleBy(2)
  println("Result of divisibleBy2(4)=" + divisibleBy2(4))
  println("Result of divisibleBy2(5)=" + divisibleBy2(5))
}

File: src/main/java/java_examples/4_functional_composition.java
package java_examples;


import java.util.function.Function;

class FunctionComposition {

    public static void main(String[] args) {
        // --------------------------------- Part 1 - Composition of a greeting ---------------------------------
        System.out.println("------------ Part 1 ------------\n");

        Function<String, String> addGreeting = name -> "Hello, my name is " + name;
        Function<String, String> toUpperCase = text -> text.toUpperCase();

        // Java supports compose and andThen to combine functions, the only difference is the order of applying
        Function<String, String> loudGreeting = addGreeting.andThen(toUpperCase);
        System.out.println(loudGreeting.apply("Tom"));


        // --------------------------------- Part 2 - Passing functions around ---------------------------------
        System.out.println("\n------------ Part 2 ------------\n");

        // First we create a generic divisible_by function
        Function<Integer, Function<Integer, Boolean>> divisibleBy = (y) -> (x) -> Math.floorMod(x, y) == 0;

        // We can now create a divisible_by_2 function using divisible_by
        // Notice divisible_by_2 is another function that returns a function
        Function<Integer, Boolean> divisibleBy2 = divisibleBy.apply(2);
        System.out.println("Result of divisibleBy2.apply(4)=" + divisibleBy2.apply(4));
        System.out.println("Result of divisibleBy2.apply(5)=" + divisibleBy2.apply(5));
    }
}
